src/components/Thumbs.js

```diff
@@ -26,81 +26,85 @@ class Thumbs extends Component {
         super(props);
 
         this.state = {
-            initialized: false,
             selectedItem: props.selectedItem,
             hasMount: false,
-            firstItem: this.getFirstItem(props.selectedItem),
-            images: []
+            firstItem: 0,
+            itemSize: null,
+            visibleItems: 0,
+            lastPosition: 0,
+            showArrows: false,
+            images: this.getImages()
         }
     }
 
-    componentDidMount (nextProps) {
-        if (!this.props.children) {
-            return;
-        }
-
+    componentDidMount(nextProps) {
         this.setupThumbs();
     }
 
-    componentWillReceiveProps (props, state) {
+    componentWillReceiveProps(props, state) {
         if (props.selectedItem !== this.state.selectedItem) {
             this.setState({
                 selectedItem: props.selectedItem,
                 firstItem: this.getFirstItem(props.selectedItem)
             });
         }
+        if (props.children !== this.props.children) {
+            this.setState({
+               images: this.getImages()
+            });
+        }
     }
 
     componentDidUpdate(prevProps) {
-        if (!prevProps.children && this.props.children && !this.state.initialized) {
-            this.setupThumbs();
+        if (this.props.children === prevProps.children) {
+            return;
         }
+
+        // This will capture any size changes for arrow adjustments etc.
+        // usually in the same render cycle so we don't see any flickers
+        this.updateSizes();
```

Comment:
When thumbs updated, re-calculate sizes, this one is AFTER render as it needs itemsWrapper clientWidth.